File: backend/app/routers/backup.py
import os
from typing import Any, Dict, List, Optional

from fastapi import APIRouter, BackgroundTasks, HTTPException
from pydantic import BaseModel

from app.services.backup_service import backup_service

router = APIRouter()


class BackupInfo(BaseModel):
    name: str
    path: str | None = None
    reason: str
    timestamp: str
    stats: dict[str, int]


class BackupResponse(BaseModel):
    success: bool
    backup_path: str | None = None
    timestamp: str | None = None
    reason: str | None = None
    stats: dict[str, int] = {}
    error: str | None = None


class RestoreResponse(BaseModel):
    success: bool
    backup_name: str | None = None
    restored_stats: dict[str, int] = {}
    previous_stats: dict[str, int] = {}
    error: str | None = None


class IntegrityResponse(BaseModel):
    healthy: bool
    current_stats: dict[str, int] = {}
    warnings: list[str] = []
    recommendations: list[str] = []


class EmergencyRecoveryResponse(BaseModel):
    success: bool
    action: str | None = None
    details: dict[str, Any] = {}
    error: str | None = None


@router.get("/list", response_model=list[BackupInfo])
async def list_backups():
    backups = backup_service.list_backups()
    return [BackupInfo(**b) for b in backups]


@router.post("/create", response_model=BackupResponse)
async def create_backup(reason: str = "manual"):
    result = backup_service.create_backup(reason)
    return BackupResponse(**result)


@router.post("/restore/{backup_name}", response_model=RestoreResponse)
async def restore_backup(backup_name: str):
    result = backup_service.restore_backup(backup_name)
    if not result["success"]:
        raise HTTPException(status_code=500, detail=result.get("error", "Restore failed"))
    return RestoreResponse(**result)


@router.get("/integrity", response_model=IntegrityResponse)
async def check_integrity():
    result = backup_service.check_data_integrity()
    return IntegrityResponse(**result)


@router.post("/emergency-recovery", response_model=EmergencyRecoveryResponse)
async def emergency_recovery():
    result = backup_service.emergency_recovery()
    return EmergencyRecoveryResponse(**result)


@router.get("/stats")
async def get_current_stats():
    from app.services.backup_service import DB_PATH
    stats = backup_service._get_db_stats(DB_PATH)
    return {
        "database": "interactive_docs.db",
        "stats": stats,
        "backup_directory": backup_service.BACKUP_DIR if hasattr(backup_service, 'BACKUP_DIR') else "backups"
    }


@router.post("/start-scheduled")
async def start_scheduled_backups(interval_hours: int = 6):
    backup_service.start_scheduled_backups(interval_hours)
    return {
        "success": True,
        "message": f"Scheduled backups started (every {interval_hours} hours)"
    }


@router.post("/stop-scheduled")
async def stop_scheduled_backups():
    backup_service.stop_scheduled_backups()
    return {
        "success": True,
        "message": "Scheduled backups stopped"
    }

File: frontend/src/api/backup.ts
import { api } from './client';

export const backupApi = {
  list: () =>
    api.get<Array<{
      name: string;
      path: string | null;
      reason: string;
      timestamp: string;
      stats: Record<string, number>;
    }>>('/backup/list'),

  create: (reason?: string) =>
    api.post<{
      success: boolean;
      backup_path: string | null;
      timestamp: string | null;
      reason: string | null;
      stats: Record<string, number>;
      error: string | null;
    }>('/backup/create', null, {
      params: { reason: reason || 'manual' }
    }),

  restore: (backupName: string) =>
    api.post<{
      success: boolean;
      backup_name: string | null;
      restored_stats: Record<string, number>;
      previous_stats: Record<string, number>;
      error: string | null;
    }>(`/backup/restore/${encodeURIComponent(backupName)}`),

  checkIntegrity: () =>
    api.get<{
      healthy: boolean;
      current_stats: Record<string, number>;
      warnings: string[];
      recommendations: string[];
    }>('/backup/integrity'),

  emergencyRecovery: () =>
    api.post<{
      success: boolean;
      action: string | null;
      details: Record<string, any>;
      error: string | null;
    }>('/backup/emergency-recovery'),

  getStats: () =>
    api.get<{
      database: string;
      stats: Record<string, number>;
      backup_directory: string;
    }>('/backup/stats'),

  startScheduled: (intervalHours?: number) =>
    api.post<{
      success: boolean;
      message: string;
    }>('/backup/start-scheduled', null, {
      params: { interval_hours: intervalHours || 6 }
    }),

  stopScheduled: () =>
    api.post<{
      success: boolean;
      message: string;
    }>('/backup/stop-scheduled'),
};
